src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs

```diff
@@ -0,0 +1,327 @@
+        }
+
+        public override bool IsInvalid => handle == IntPtr.Zero;
+    }
+
+    internal class SafeKeychainHandle : SafeHandle
+    {
+        internal SafeKeychainHandle()
+            : base(IntPtr.Zero, ownsHandle: true)
+        {
+        }
+
+        internal SafeKeychainHandle(IntPtr handle)
+            : base(handle, ownsHandle: true)
+        {
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.CoreFoundation.CFRelease(handle);
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+
+        public override bool IsInvalid => handle == IntPtr.Zero;
+    }
+
+    internal sealed class SafeTemporaryKeychainHandle : SafeKeychainHandle
+    {
+        private static readonly Dictionary<IntPtr, SafeTemporaryKeychainHandle> s_lookup =
+            new Dictionary<IntPtr, SafeTemporaryKeychainHandle>();
+
+        internal SafeTemporaryKeychainHandle()
+        {
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            lock (s_lookup)
+            {
+                s_lookup.Remove(handle);
+            }
+
+            Interop.AppleCrypto.SecKeychainDelete(handle, throwOnError: false);
+            return base.ReleaseHandle();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && SafeHandleCache<SafeTemporaryKeychainHandle>.IsCachedInvalidHandle(this))
+            {
+                return;
+            }
+
+            base.Dispose(disposing);
+        }
+
+        public static SafeTemporaryKeychainHandle InvalidHandle =>
+            SafeHandleCache<SafeTemporaryKeychainHandle>.GetInvalidHandle(() => new SafeTemporaryKeychainHandle());
+
+        internal static void TrackKeychain(SafeTemporaryKeychainHandle toTrack)
+        {
+            if (toTrack.IsInvalid)
+            {
+                return;
+            }
+
+            lock (s_lookup)
+            {
+                Debug.Assert(!s_lookup.ContainsKey(toTrack.handle));
+
+                s_lookup[toTrack.handle] = toTrack;
+            }
+        }
+
+        internal static void TrackItem(SafeKeychainItemHandle keychainItem)
+        {
+            if (keychainItem.IsInvalid)
+                return;
+
+            using (SafeKeychainHandle keychain = Interop.AppleCrypto.SecKeychainItemCopyKeychain(keychainItem))
+            {
+                if (keychain.IsInvalid)
+                {
+                    return;
+                }
+
+                lock (s_lookup)
+                {
+                    SafeTemporaryKeychainHandle temporaryHandle;
+
+                    if (s_lookup.TryGetValue(keychain.DangerousGetHandle(), out temporaryHandle))
+                    {
+                        bool ignored = false;
+                        temporaryHandle.DangerousAddRef(ref ignored);
+                    }
+                }
+            }
+        }
+
+        internal static void ReleaseItem(IntPtr keychainItem)
```

Comment:
Fair enough.  It was "ReleaseItem" since it got called from "ReleaseHandle", and itself calls DangerousRelease; but "UntrackItem" is a better parallel to "TrackItem".  And makes it more clear that I'm not (necessarily) freeing anything.